rpc/kate-rpc/src/lib.rs

```diff
@@ -141,6 +148,77 @@ where
 	SDFilter: Filter<CallOf<Block>> + 'static,
 	<<Block as BlockT>::Extrinsic as Extrinsic>::Call: Clone,
 {
+	fn query_rows(
+		&self,
+		rows: Vec<u32>,
+		at: Option<HashOf<Block>>,
+	) -> RpcResult<Vec<Option<Vec<u8>>>> {
+		let block_id = self.block_id(at);
+
+		let signed_block = self
+			.client
+			.block(&block_id)
+			.map_err(|e| internal_err!("Invalid block number: {:?}", e))?
+			.ok_or_else(|| internal_err!("Missing block {}", block_id))?;
+		let block_hash = signed_block.block.header().hash();
+
+		let mut block_ext_cache = self
+			.block_ext_cache
+			.write()
+			.map_err(|_| internal_err!("Block cache lock is poisoned .qed"))?;
+
+		let block_length: BlockLength = self
+			.client
+			.runtime_api()
+			.get_block_length(&block_id)
+			.map_err(|e| internal_err!("Block Length cannot be fetched: {:?}", e))?;
+
+		if !block_ext_cache.contains(&block_hash) {
+			// build block data extension and cache it
+			let xts_by_id: Vec<AppExtrinsic> = signed_block
+				.block
+				.extrinsics()
+				.iter()
+				.map(|e| AppExtrinsic {
+					app_id: e.app_id(),
+					data: e.encode(),
+				})
+				.collect();
+
+			// Use Babe's VRF
+			let seed: [u8; 32] = self
+				.client
+				.runtime_api()
+				.get_babe_vrf(&block_id)
+				.map_err(|e| internal_err!("Babe VRF not found for block {}: {:?}", block_id, e))?;
+
+			let (_, block, block_dims) = kate::com::flatten_and_pad_block(
+				block_length.rows,
+				block_length.cols,
+				block_length.chunk_size(),
+				&xts_by_id,
+				seed,
+			)
+			.map_err(|e| internal_err!("Flatten and pad block failed: {:?}", e))?;
+
+			let data = kate::com::par_extend_data_matrix(block_dims, &block)
+				.map_err(|e| internal_err!("Matrix cannot be extended: {:?}", e))?;
+
+			block_ext_cache.put(block_hash, (data, block_dims));
```

Comment:
QQ: Can we insert block in cache if and only if that block is finalized?

Reply:
How to check if block is finalized?